File: 229_majority_element_II.cpp
class Solution {
public:
    vector<int> majorityElement(vector<int>& nums) {
    int n1,n2;
    int c1=0,c2=0;
        for (int i=0;i<nums.size();i++) {
            if (nums[i]==n1) c1++;
            else if (nums[i]==n2) c2++;
            else if (!c1) {n1=nums[i];c1++;}
            else if (!c2) {n2=nums[i];c2++;}
            else {c1--;c2--;}
        }
        c1=0;c2=0;
        for (int i=0;i<nums.size();i++) {
            if (nums[i]==n1) c1++;
            else if (nums[i]==n2) c2++;
        }
        vector<int> result;
        if (c1>nums.size()/3) result.push_back(n1);
        if (c2>nums.size()/3) result.push_back(n2);
        return result;
            
    }
};

File: algorithms/python3/229_majority_element_II.py
class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        c1=c2=0
        a1=a2=0
        for num in nums:
            if num==a1:
                c1+=1
            elif num==a2:
                c2+=1
            elif c1==0:
                a1=num
                c1=1
            elif c2==0:
                a2=num
                c2=1
            else:
                c1-=1
                c2-=1
        c1=c2=0
        for num in nums:
            if num==a1:
                c1+=1
            elif num==a2:
                c2+=1
        ans=[]
        if c1>len(nums)//3:
            ans.append(a1)
        if c2>len(nums)//3:
            ans.append(a2)
        return ans


a=Solution()

print(a.majorityElement([1,1,1,3,3,2,2,2]))